Add private docker registry support

Egor-S opened the pull request:
Closes #246

* Added `registry_auth` for `docker ` provider
* Variables interpolation by the runner
  * `$VAR_NAME`
  * `${VAR_NAME}` and `${ VAR_NAME }`
  * `$$` for escaped `$`
  * Variables come from secrets
  * Warning if a variable is missed
  * Only `registry_auth.username` and `registry_auth.password` at this moment
* Interpolation string validation by the CLI
* Tests for interpolation and validation
* Docs update

Egor-S committed:
Pull images from private registry without secrets interpolation

Egor-S committed:
Python interpolation string validator

Egor-S committed:
Extend interpolation tests

Egor-S committed:
Go variable interpolation

Egor-S committed:
Test go interpolation

Egor-S committed:
Test registry_auth for interpolation errors

Egor-S committed:
close div tag: environment-variables.md

Egor-S committed:
Add docs for registry_auth

r4victor commented on docs/usage/private-docker-registry.md:
This page fits nowhere in the docs structure. I would suggest to not create a separate page to describe Private Docker Registry. We can describe `registry_auth` property on the docker reference page as we do for other properties.

<img width="1510" alt="Screenshot 2023-03-20 at 14 05 16" src="https://user-images.githubusercontent.com/16836845/226293736-38de2fdb-5d04-43d9-9b1a-9689caf6f4f7.png">

r4victor commented on cli/dstack/core/interpolation.py:
`interpolation.py` doesn't feel like it belongs to `core` directory because all other modules there are centered around data models. I'd move the interpolation logic to `job.py` or some other model file or to a separate file in `utils`.

Egor-S replied:
Will move to utils

Egor-S committed:
Add private-docker-registry.md to side menu

Egor-S committed:
JinjaMock for preventing secrets interpolation

Egor-S committed:
Rewrite go interpolator & tests; update docs

Egor-S commented:
Reworked interpolation syntax:

* `${{ secrets.VAR_NAME }}` used for secrets
  * `$$` escapes `$`, but it is really needed only before `{{`
* `JinjaMock` prevents client-side interpolation
* Updated tests
* Client-side validation is handled by jinja2 template rendering
* Fixed docs

r4victor commented on cli/dstack/utils/jinjamock.py:
```diff
@@ -0,0 +1,27 @@
+        self.__n__ = namespace
+
+    def __getattr__(self, item: str) -> JinjaMockValue:
+        namespace = self.__dict__["__n__"]
```
Seem like some weird magic. Why can't we just store `self.namespace = namespace`? Why access via self.__dict__? Why return `JinjaMockValue`? Can't we just return a string? Why name variables like __s__ – i.e. double underscores on both sides usually means it has special meaning to the interpreter.

Not critical, just reads weird.

Egor-S replied:
I tried to invent something unlikely to be used as a secret name

`${{ secrets.namespace }}` is more likely than `${{ secrets.__n__ }}` in my opinion, but brokes `JinjaMock`

`JinjaMockValue` is designed to raise errors in indexing and chaining. `str` would allow indexing and break `JinjaMock`: `${{ secrets.password[1] }}` and `${{ secrets.password.hash }}`

Probably I could avoid using `__dict__` there

Egor-S replied:
`JinjaMock` is removed

Egor-S committed:
Replace jinja2 interpolation with simplified syntax; rewrite tests

r4victor commented on cli/dstack/utils/interpolator.py:
```diff
@@ -0,0 +1,68 @@
+            if s[dollar + 1] == "$":  # escaped $$
+                tokens.append("$")
+                start += 2
+                continue
```
Seems like the code doesn't handle escaped $$ in the middle of the string. E.g. `self.assertEqual("echo ${{ENV}}", get_interpolator().interpolate("echo $${{ENV}}"))` fails.

Egor-S committed:
Fix dollar escaping; add test
